sim/firesim-lib/src/main/scala/bridges/TerminationBridge.scala

```diff
@@ -0,0 +1,136 @@
+  //It is just inputs from the target to indicate test completion
+  //The HostPort annotation would not work with aggregates so creating individual channels.
+  val errCodeH = InputChannel(targetIO.errCode)
+  val validH = InputChannel(targetIO.valid)
+}
+
+class TerminationBridge(params: TerminationBridgeParams) extends BlackBox
+    with Bridge[TerminationBridgeHostIO, TerminationBridgeModule] {
+  val io = IO(new TerminationBridgeTargetIO(params))
+  val bridgeIO = new TerminationBridgeHostIO(params)(io)
+
+  val constructorArg = Some(params)
+
+  generateAnnotations()
+}
+
+object TerminationBridge {
+
+  private def annotateTerminationBridge(clock: Clock, lReset: Reset, finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
+    val finishAll = finish.reduce(_||_)
+    val errMessageID = PriorityEncoder(finish)
+    //val gReset = WireInit(false.B)
+    //GlobalResetCondition.sink(gReset)
+    val terminationBridgeTarget = Module(new TerminationBridge(params))
+    terminationBridgeTarget.io.errCode := errMessageID
+    terminationBridgeTarget.io.valid := finishAll && !lReset.asBool
+    terminationBridgeTarget.io.clock := clock 
+    terminationBridgeTarget  
+  }
+
+  def apply(clock: Clock, lReset: Reset, finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
+    annotateTerminationBridge(clock, lReset, finish, errCode, params)
+  }
+
+  def apply(clock: Clock, lReset: Reset, finish: Bool, errCode: UInt, params: TerminationBridgeParams): TerminationBridge = {
+    //val gReset = WireInit(false.B)
+    //GlobalResetCondition.sink(gReset)
+    val terminationBridgeTarget = Module(new TerminationBridge(params))
+    terminationBridgeTarget.io.errCode := errCode
+    terminationBridgeTarget.io.valid := finish && !lReset.asBool
+    terminationBridgeTarget.io.clock := clock 
+    terminationBridgeTarget  
+  }
+
+  def apply(finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
+    annotateTerminationBridge(Module.clock, Module.reset, finish, errCode, params)
+  }
+
+}
+
+class TerminationBridgeModule(params: TerminationBridgeParams)(implicit p: Parameters) extends BridgeModule[TerminationBridgeHostIO]()(p) {
+  lazy val module = new BridgeModuleImp(this) {
+    // This creates the interfaces for all of the host-side transport
+    // AXI4-lite for the simulation control bus, =
+    // AXI4 for DMA
+    val io = IO(new WidgetIO())
+
+
+
+    // This creates the host-side interface of your TargetIO
+    val hPort = IO(new TerminationBridgeHostIO(params)())
+    
+    //Resume the simulation after the target called for termination
+    val resume = WireInit(false.B)
+
+    val statusDone = Module(new Queue(Bool(), 1))
+    val errCode = Module(new Queue(hPort.errCodeH.bits.cloneType, 1))
+
+    val cycleCountWidth = 16 //Data width is capped by nastikey 
```

Comment:
You should use a 64b counter for this. (The width is actually capped at 32b). 

Here's an example of how that looks: https://github.com/firesim/firesim/blob/366ceabe13ce27056ec1c53bf93d017ec182d85e/sim/midas/src/main/scala/midas/widgets/ClockBridge.scala#L172

Reply:
And here's the C side: https://github.com/firesim/firesim/blob/366ceabe13ce27056ec1c53bf93d017ec182d85e/sim/midas/src/main/cc/simif.cc#L56

Notice the write to "latch" which provides a hacky way to do an "atomic" 64b read using two 32b reads.

Reply:
All three of those addresses (_h, _l, _latch) will be autogenerated by the scala above.

Reply:
Done